### PR TITLE
Additional processes

### PR DESCRIPTION
**Pull request by claxn**
- simple mathematical processes are implemented
- extended functionality for exporting images

### CODE REVIEW

File: src/processes/exp.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const Process = require('../processgraph/process');
+
+module.exports = class exp extends Process {
+
+    async execute(node, context) {
```
Comment (m-mohr) [inline src/processes/exp.js]: Does this work as expected? Not sure where we get the numbers from...

File: src/processes/first.json [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  "id": "first",
+  "summary": "First element",
+  "description": "Gives the first element of an array. For an empty array `null` is returned.",
+  "categories": [
+    "arrays",
+    "reducer"
+  ],
+  "parameter_order": [
+    "data",
+    "ignore_nodata"
```
Comment (m-mohr) [inline src/processes/first.json]: You forgot to remove this.

File: src/processes/first.json [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+      "name": "data"
+    }
+  },
+  "returns": {
+    "description": "The first element of the input array.",
+    "schema": {
+      "description": "Any data type is allowed."
+    }
+  },
+  "examples": [
```
Comment (m-mohr) [inline src/processes/first.json]: The examples must be checked if there's no ignore_nodata parameter. At least the third one seems incorrect.

File: src/processes/linear_scale_range.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        var inputMax = node.getArgument('inputMax');
+        var outputMin = node.getArgument('outputMin');
+        var outputMax = node.getArgument('outputMax');
+        if (outputMin == null){
+            outputMin = 0
+        }
+        if (outputMax == null){
+            outputMax = 1
+        }
+        var process = function(image){
```
Comment (m-mohr) [inline src/processes/linear_scale_range.js]: I think it would be better style to define this function in the class directly.

File: src/processes/ndvi.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+const Process = require('../processgraph/process');
+
+// TODO: assumes bands have been labelled before
```
Comment (m-mohr) [inline src/processes/ndvi.js]: This is not how the NDVI function is planned to work. It should check the metadata for the common names...

File: src/processes/power.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+const Process = require('../processgraph/process');
+const Commons = require('../processgraph/commons');
+
+// TODO test implementation
```
Comment (m-mohr) [inline src/processes/power.js]: That would be good ;-)

File: src/processes/quantiles.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+const Process = require('../processgraph/process');
+const Commons = require('../processgraph/commons');
+const Errors = require('../errors');
+const underscore = require('underscore');
+
+// TODO: should we check for invalid input, e.g. probabilites > 1?
```
Comment (m-mohr) [inline src/processes/quantiles.js]: Yes, why not? But in validate, not execute. - or can we add such things to the JSON schema?

File: src/processes/round.json [hunk 8]
```diff
@@ -0,0 +1,84 @@
+{
+  "id": "round",
+  "summary": "Rounds to a specified precision",
+  "description": "Rounds a real number `x` to specified precision `p`.\n\nIf the fractional part of `x` is halfway between two integers, one of which is even and the other odd, then the even number is returned.\nThis behaviour follows [IEEE Standard 754](https://ieeexplore.ieee.org/document/4610935). This kind of rounding is also called \"rounding to nearest\" or \"banker's rounding\". It minimizes rounding errors that result from consistently rounding a midpoint value in a single direction.\n\nThe no-data value `null` is passed through and therefore gets propagated.",
+  "categories": [
+    "math > rounding"
+  ],
+  "parameter_order": [
```
Comment (m-mohr) [inline src/processes/round.json]: You removed p, but it's still mentioned in the description and examples. I think we can simply implement p ourself if it's not supported by Google directly. Just with some math...

File: src/processes/sd.json [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+  "id": "sd",
+  "summary": "Standard deviation",
+  "description": "Computes the sample standard deviation, which quantifies the amount of variation of an array of numbers. It is defined to be the square root of the corresponding variance (see ``variance()``).\n\nA low standard deviation indicates that the values tend to be close to the expected value, while a high standard deviation indicates that the values are spread out over a wider range.",
+  "categories": [
+    "math",
+    "reducer"
+  ],
+  "parameter_order": [
+    "data",
+    "ignore_nodata"
```
Comment (m-mohr) [inline src/processes/sd.json]: ignore_nodata still in here without having a parameter, also examples need to be checked.

File: src/processes/sort.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+const Process = require('../processgraph/process');
+const Commons = require('../processgraph/commons');
+
+// TODO: complete implementation
+module.exports = class sort extends Process {
+
+    async execute(node, context) {
```
Comment (m-mohr) [inline src/processes/sort.js]: Does this sort works as promised by the schema? Does it allow dates/times/strings and all that? Does it work on images or arrays? Is this tested?

File: src/processes/sum.json [hunk 11]
```diff
@@ -0,0 +1,86 @@
+{
+  "id": "sum",
+  "summary": "Addition of a sequence of numbers",
```
Comment (m-mohr) [inline src/processes/sum.json]: ignore_nodata is still mentioned here. Have you tested how sum works in GEE with no data values?

File: src/processes/variance.json [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+  "id": "variance",
+  "summary": "Variance",
+  "description": "Computes the sample variance of an array of numbers by calculating the square of the standard deviation (see ``sd()``). It is defined to be the expectation of the squared deviation of a random variable from its expected value. Basically, it measures how far the numbers in the array are spread out from their average value.",
+  "categories": [
+    "math",
+    "reducer"
+  ],
+  "parameter_order": [
+    "data",
+    "ignore_nodata"
```
Comment (m-mohr) [inline src/processes/variance.json]: ignore_nodata issue again... ;-)

File: src/processgraph/commons.js [hunk 13]
```diff
@@ -3,17 +3,42 @@ const Utils = require('../utils');
...
 module.exports = class ProcessCommons {
 
-	static reduceInCallback(node, reducer) {
+	static reduceInCallback(node, reducer, ...args) {
 		var dc = node.getData("data");
 		var func = data => data.reduce(reducer);
 		if (node.getProcessGraph().isSimpleReducer() || dc.isImageCollection()) {
-			dc.imageCollection(func);
+			dc.imageCollection(func, ...args);
+			// rename bands according to the GEE convention
```
Comment (m-mohr) [inline src/processgraph/commons.js]: Is this renaming backed by the process spec? The band dimension should be removed, I think. We should have a way to work with it on Googles side. So maybe it's better to rename the bands on Googles side to a fixed name that indicates us there's no band left? Not sure myself, but needs discussion...

File: src/processgraph/context.js [hunk 14]
```diff
@@ -87,14 +112,21 @@ module.exports = class ProcessingContext {
...
-		else {
+		else if(format.toLowerCase() === 'json') {
 			var fileName = Utils.generateHash() + "/result-" + Date.now() +  "." + this.translateOutputFormat(format);
 			var p = path.normalize(path.join(this.serverContext.getTempFolder(), fileName));
 			var parent = path.dirname(p);
 			await fse.ensureDir(parent);
 			await fse.writeJson(p, dataCube.getData());
 			return Utils.getApiUrl("/temp/" + fileName);
 		}
+		else{
```
Comment (m-mohr) [inline src/processgraph/context.js]: That should never be the case as it should be catched in validation. So this can be removed then...

In the best case we never throw errors in execution as this costs money and/or resources. Most errors should be thrown in validation.

File: src/processgraph/datacube.js [hunk 15]
```diff
@@ -29,91 +29,99 @@ module.exports = class DataCube {
...
-	isImage() {
-		if (this.data instanceof ee.Image) {
-			return true;
+	computedObjectType(){
+		return this.data.getInfo().type
+	}
+
+	objectType(){
+		if (this.data instanceof ee.Image){
+			return "Image"
```
Comment (m-mohr) [inline src/processgraph/datacube.js]: Some semicolons missing in this function ;-)

File: src/processgraph/datacube.js [hunk 16]
```diff
@@ -217,7 +225,7 @@ module.exports = class DataCube {
 	}
 
 	getBands() {
-		return this.dimBands().getValues();
+		return this.dimBands().values;
```
Comment (m-mohr) [inline src/processgraph/datacube.js]: Why this change?

File: src/processes/apply.js [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    async execute(node, context) {
+        var dc = node.getData("data");
+        var callback = node.getArgument("process");
+        // TODO: there is for sure a better way to do this
+        // TODO: outsource it to commons.js
+        var arg_dict = null;
+        if(callback.resultNode.arguments.hasOwnProperty('x')){
+            arg_dict = {x: dc};
+        }
+        else{  // this assumes that the argument names have been checked before
```
[the remark this answers is not in the document]
Comment (claxn) [reply src/processes/apply.js]: Some processes use 'x' as a argument, some use 'data', which causes problems with the callback.

[the remark this answers is not in the document]
Comment (m-mohr) [reply src/processes/apply.js]: But why's that a problem? The name of the parameter shouldn't be an issue, I guess? It is always the first argument and therefore the name should be irrelevant? Do I miss something here?

File: src/processes/cosh.js [hunk 18]
```diff
@@ -0,0 +1,13 @@
+const Process = require('../processgraph/process');
+const Commons = require('../processgraph/commons');
+
+module.exports = class cosh extends Process {
+
+    async execute(node, context) {
+        var image_process = function(image){
```
[the remark this answers is not in the document]
Comment (claxn) [reply src/processes/cosh.js]: You are right, way better, I will do so!

File: src/processgraph/datacube.js (hunk 16 above)
Comment (claxn) [reply src/processgraph/datacube.js]: Mmh, strange, now it works again with getValues(). Before I always got an error using it. I have reverted my change

File: src/processgraph/context.js (hunk 14 above)
Comment (claxn) [reply src/processgraph/context.js]: Okay, I will keep this in my mind!

File: src/processgraph/commons.js (hunk 13 above)
Comment (claxn) [reply src/processgraph/commons.js]: Currently we are always reducing over time, since this is the only way to do it for image collections. However, if we convert them to arrays we could resolve this and also reduce along the band dimension. For now I would just rename them back to their original name (e.g. 'VV_mean' -> 'VV'). Then the user can access the bands as intended, without any knowledge about the GEE naming convention.

File: src/processgraph/datacube.js (hunk 16 above)
Comment (claxn) [reply src/processgraph/datacube.js]: I applied the changes with the GEE naming convention and now this error pops up again:

`TypeError: this.dimBands(...).getValues is not a function`

Using values as before works fine.

Comment (m-mohr) [reply src/processgraph/datacube.js]: Okay, then let's try to fix that. I guess dimBands() returns null if there's no band dimension and then it fails to call getValues on null. Can you confirm which value is returned by dimBands()?

File: src/processgraph/commons.js (hunk 13 above)
Comment (m-mohr) [reply src/processgraph/commons.js]: That sounds reasonable.

File: src/processes/sort.js (hunk 10 above)
Comment (claxn) [reply src/processes/sort.js]: No, I will remove it for the time being until we have a solution how we can properly work on arrays.

File: src/processes/sum.json (hunk 11 above)
Comment (claxn) [reply src/processes/sum.json]: I forgot to remove some parts where `ignore_data` is mentioned. GEE just ignores them so it is always true. We have to ask Google if can somehow get a control over nodata values.

Comment (m-mohr) [reply src/processes/sum.json]: I doubt we can anytime soon.

File: src/processgraph/datacube.js (hunk 16 above)
Comment (claxn) [reply src/processgraph/datacube.js]: image

Comment (m-mohr) [reply src/processgraph/datacube.js]: Oh, wait. Why's there no getValues() in the Dimension class? Seems I missed to implement it. Could you simply add one? :-)

File: src/processes/ndvi.js (hunk 5 above)
Comment (claxn) [reply src/processes/ndvi.js]: Do you mean one has to check the STAC metadata for appropriate band synonyms?

Comment (m-mohr) [reply src/processes/ndvi.js]: Yes, the back-end should easily be able to do that.
